test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +462,288 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
+// successfully predecoded and that more data is fetched subsequently allowing
+// progression as a full instruction is now present in the buffer
+TEST_P(PipelineFetchUnitTest, invalidMinBytesAtEndOfBuffer) {
+  // This is only relevant if min and max size are different. Otherwise, there
+  // won't be any progression as the fetch unit will be caught in an infinite
+  // loop
+  if (insnMinSizeBytes < insnMaxSizeBytes) {
+    ON_CALL(isa, getMaxInstructionSize())
+        .WillByDefault(Return(insnMaxSizeBytes));
+    ON_CALL(isa, getMinInstructionSize())
+        .WillByDefault(Return(insnMinSizeBytes));
+    ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+    // Buffer will contain invalid min bytes so predecode returns 0 bytes read
+    ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+        .WillByDefault(Return(0));
+
+    // getMaxInstructionSize called for second time in assertion
+    if (strcmp(SIMENG_BUILD_TYPE, "Release") == 0) {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    } else {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(2);
+    }
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    uint64_t setPC = blockSize - insnMinSizeBytes;
+    // Fetch a single minimum sized instruction, buffered bytes = 0
+    fetchUnit.updatePC(setPC);
+    // Tick
+    fetchUnit.tick();
+
+    // No data consumed
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], MacroOp());
+
+    // Expect that memory is requested even though there is data in the buffer
+    // as bufferedBytes < maxInstructionSize
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory,
+                requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+        .Times(1);
+    fetchUnit.requestFromPC();
+
+    // Tick again expecting buffer to be filled and a word is predecoded
+    MacroOp mOp = {uopPtr};
+    memory::MemoryReadResult nextBlockValue = {{16, blockSize}, 0, 1};
+    span<memory::MemoryReadResult> nextBlock = {&nextBlockValue, 1};
+    ON_CALL(memory, getCompletedReads()).WillByDefault(Return(nextBlock));
+    ON_CALL(isa, predecode(_, _, _, _))
+        .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(insnMaxSizeBytes)));
+
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    fetchUnit.tick();
+
+    // Initially min bytes, 16 bytes added, max bytes predecoded
+    EXPECT_EQ(fetchUnit.bufferedBytes_,
+              (insnMinSizeBytes + 16) - insnMaxSizeBytes);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp);
```

Comment:
I think rather than address 16, we should base the next fetch address off of the block size